### ISSUE TITLE
Multi-node training without AWS EFA clusters

### ISSUE THREAD

**Issue by LeoXinhaoLee** [pytorch/torchtitan#566]
Thank you so much for releasing code for this great project!

For multi-node training, right now I've only found commands in `multinode_trainer.slurm`, which seem to be specific to AWS EFA slurm clusters.

I'm wondering if it's possible to try multi-node training without ASW setup, say with simply the IPs of 2 nodes instead?

Thank you very much for your help!

**Comment by lessw2020** [comment]
Hi @LeoXinhaoLee  - thanks for raising this issue.
The short answer to your question is yes you can setup to run directly with two nodes.
The reason this isn't shown is that the concern would be you are (likely) running over ethernet to do this and the perf would be very bad.

But the basics to do this if you are planning to run smaller models and/or have a faster connection would be to run a slightly modified torch launch on the master node vs the compute nodes. 

~~~
--node_rank=0. 
~~~
for the master node and then 1 etc. for the other nodes.  You'd also pass in the master_addr to make sure they can all link up. 

This type of setup was more common for DDP, where the between nodes comms would be lighter (just the gradients for all_reduce) rather than with FSDP where you are sending weights and optimizer states as well. 
Note - as I write that if you are interested in trying PP, then that would be your lightest weight comms across nodes...so actually that would make this more feasible. 

There's an older article that talks about doing this along with using mpirun so you don't have to manually run torchrun by hand on each node:
https://lambdalabs.com/blog/multi-node-pytorch-distributed-training-guide#launch-multi-node-pytorch-distributed-applications

Anyway, I'd recommend trying with PP as the lightest across node comms with titan.
Hope that helps and feel free to post if you hit issues or have more questions when trying it out!

**Comment by lessw2020** [comment]
Hi @LeoXinhaoLee  - 
1 - Yes, if you have infiband or any similar high speed setup between nodes, then you should be in good shape! 
2 - Titan uses FSDP2, and we do have a PR to add HSDP support landing in the next week or two.  The PR has some good info as well on how to use it once it lands:
https://github.com/pytorch/torchtitan/pull/518

HSDP would also greatly reduce the cross node comms, so this could be a nice option as soon as this lands.

**Comment by LeoXinhaoLee** [comment]
I see, looking forward to seeing HSDP. Thank you very much!
